### DISCUSSION TITLE
Reading full 32-bit data stream on sink

### DISCUSSION THREAD

**Discussion by riskray** [pschatzmann/ESP32-A2DP#183] [Q&A] [answered]
Hi, I’m trying to access the full data stream for a sink to do some signal processing on it, but using set_stream_reader only allows 8 bits of the data, resulting in clipping, I was wondering if it’s possible to access the whole stream or remove the clipping by converting the 32 but stream to 8 bits.

thanks in advance

**Comment by pschatzmann** [comment, marked as answer]
No, it's not 8 bits. Its int16_t with 2 channels...

**Reply by riskray** [reply]
Sorry maybe I wasn't clear, in your README, there is an example for accessing the data packets like this:
```
a2dp_sink.set_stream_reader(read_data_stream);
void read_data_stream(const uint8_t *data, uint32_t length)
{
  // Do something with the data packet
}
```
And the set_stream_reader function takes uint8_t as a parameter, only 8-bits. Although it seems as I've looked at them its converted from a larger number therefore cutting off more significant bits.
Thanks again in advance.

**Reply by pschatzmann** [reply]
I agree, the definition is a bit unfortunate: You receive a byte stream. But the underlying data is int16[2][]
Just can cast it into an int16_t*. and take even values as left and odd values as right channel...
The number of samples is len/2
The number of frames is len/4

The definition has been done this way because I usually just write the data to an Arduino Stream...

**Reply by riskray** [reply]
Thanks a lot, this has helped

**Reply by AvishkaDon** [reply]
Hi, is there a way to write this data (some of it will be processed) back to a 12S stream that goes to an external DAC?

**Reply by pschatzmann** [reply]
I suggest that you use my arduino audio tools library for this. Have a look at [this example](https://github.com/pschatzmann/arduino-audio-tools/blob/main/examples/examples-basic-api/basic-a2dp-i2s/basic-a2dp-i2s.ino)

**Comment by pschatzmann** [comment]
I have updated the readme to be a little bit clearer on this...
